### DISCUSSION TITLE
Is this m1 architecture issue?

### DISCUSSION THREAD

**Discussion by yoonsik-jung-opt** [alandefreitas/matplotplusplus#256] [Q&A]
I just try to run the example codes below. 

```c++
#include <random>

int main() {
    using namespace matplot;

    auto x = linspace(0, 3 * pi, 200);
//    auto y = transform(x, [&](double x) { return cos(x) + rand(0, 1); });

//    scatter(x, y);

//    show();
    return 0;
}
```

After run the code on CLion, I got the error message. I installed the matplot++ using brew and link to my project.
I guess my chip architecture of macbook causes the error, but I am not sure. (I am newbie on c++) . 

There are logmsg and cmakelist.txt that I used. 
Thanks.

```
cmake_minimum_required(VERSION 3.21)
project(LADMH)
set(CMAKE_CXX_STANDARD 14)
set(Matplot++_DIR "")
include_directories(/usr/local)
find_package(Matplot++ REQUIRED)

if (Matplot++_FOUND)
    add_executable(LADMH main.cpp preprocessing.h pattern_generation.h utilities.h)
    target_link_libraries(LADMH PUBLIC Matplot++::matplot)
endif ()

```
**Output**

<details>
 
```console
[1/1] Linking CXX executable LADMH
FAILED: LADMH 
: && /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -g -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names  CMakeFiles/LADMH.dir/main.cpp.o -o LADMH  -Wl,-rpath,/usr/local/lib  /usr/local/lib/libmatplot.dylib && :
ld: warning: ignoring file /usr/local/lib/libmatplot.dylib, building for macOS-arm64 but attempting to link with file built for macOS-x86_64
Undefined symbols for architecture arm64:
  "matplot::linspace(double, double, unsigned long)", referenced from:
      _main in main.cpp.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```
</details>

**Steps to Reproduce**

```bash
# Your steps go here
/Applications/CLion.app/Contents/bin/cmake/mac/bin/cmake --build /Users/myname/Desktop/workspace/PRJNAME/cmake-build-debug --target PRJNAME
```

**Platform**

- [ ] *cross-platform issue - linux*
- [ ] *cross-platform issue - windows*
- [x] *cross-platform issue - macos*

**Environment Details:**
 - OS: Monterey(Apple M1 Pro)
 - OS Version: 12.3.1
 - Compiler:clang
 - Compiler version:12.0.5

**Additional context**

**Comment by alandefreitas** [comment]
In your example you have

```cpp
#include <random>
```

but still need the matplot headers.

I don't know if the version from brew works OK on arm processors. You can always install from source though.

In your CMake, this

```cmake
set(Matplot++_DIR "")
```

and this

```cmake
include_directories(/usr/local)
```

are very unusual.

- The find_package script is supposed to set `Matplot++_DIR` for you. 
- Nowadays we use `target_include_directories` instead of `include_directories`
- And there's no point in including the whole `/usr/local/` directory, as you'll get the header files but not the complete library targets you need to link.

Also, if you have `REQUIRED` in your

```cmake
find_package(Matplot++ REQUIRED)
```

then you don't need

```cmake
if (Matplot++_FOUND)
```

because `REQUIRED` will already fail when Matplot++ is not found (when `NOT Matplot++_FOUND`).

Then your error is `ignoring file /usr/local/lib/libmatplot.dylib` because the library you installed is `macOS-x86_64` and your computer is `macOS-arm64`. `x86_64` is the version you get from the release section. I don't know what the brew recipe does here. But you can always 1) install from source for `arm64` and `find_package`, 2) clone the repository and add matplot++ as a CMake subdirectory, or 3) use CMake FetchContent.

**Reply by yoonsik-jung-opt** [reply]
Thank you for your kindness answer! but unfortunately, I still not clear some points, then have a few questions.

I uninstalled the matplotplusplus via brew and changed my CMakeList.txt as using CMake subdirectory(I referenced https://alandefreitas.github.io/matplotplusplus/integration/cmake/embed-as-subdirectory/). It not only works for import the library, however, but also raises errors. (In CLion, it shows parameters of functions in matplot++, after successfully loaded matplot/matplot.h)

```
cmake_minimum_required(VERSION 3.21)
project(LADMH)
set(CMAKE_CXX_STANDARD 14)
add_subdirectory(matplotplusplus)

add_executable(LADMH main.cpp preprocessing.h pattern_generation.h utilities.h)
target_link_libraries(LADMH PUBLIC matplot)
```

```
====================[ Build | LADMH | Debug ]===================================
/Applications/CLion.app/Contents/bin/cmake/mac/bin/cmake --build /Users/yoonsikjung/Desktop/workspace/LADMH/cmake-build-debug --target LADMH
[2/3] Building CXX object CMakeFiles/LADMH.dir/main.cpp.o
FAILED: CMakeFiles/LADMH.dir/main.cpp.o 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -DMATPLOT_BUILD_HIGH_RESOLUTION_WORLD_MAP -DNOMINMAX -I/Users/yoonsikjung/Desktop/workspace/LADMH/matplotplusplus/source -g -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk -std=gnu++17 -MD -MT CMakeFiles/LADMH.dir/main.cpp.o -MF CMakeFiles/LADMH.dir/main.cpp.o.d -o CMakeFiles/LADMH.dir/main.cpp.o -c /Users/yoonsikjung/Desktop/workspace/LADMH/main.cpp
/Users/yoonsikjung/Desktop/workspace/LADMH/main.cpp:6:10: fatal error: 'CImg.h' file not found
#include "CImg.h"
         ^~~~~~~~
1 error generated.
ninja: build stopped: subcommand failed.
```

I tried to install cimg via brew, subdirectory, and source, but all tries are failed....How can I do for successfully load the CImg.h?

**Reply by alandefreitas** [reply]
Short answer: The problem here is you're attempting to include "CImg.h" in your program. But you didn't link the `cimg` library to your target. You might think linking to `matplot` would be enough, but `cimg` is a private dependency of `matplot`, which is not exposed to targets linking `matplot`. You should link the `cimg` target created by the matplot scripts or you could directly include the directory with `cimg` in your own target, depending on how you're using it.

The relevant script that creates the cimg target is here:

https://github.com/alandefreitas/matplotplusplus/blob/8e1ca55d53ce35d8961ea1c776279fdb04ac4457/source/3rd_party/CMakeLists.txt#L54

Then it links that dependency with matplot as `PRIVATE`:

https://github.com/alandefreitas/matplotplusplus/blob/8e1ca55d53ce35d8961ea1c776279fdb04ac4457/source/matplot/CMakeLists.txt#L104

It sets the `CIMG_INCLUDE_DIR` variable and includes the target. As a starting point, you could output that and check if this matches the appropriate CImg directory. IF the `cimg` target is available, you can link that with your own target. At worst, you could use `target_include_directories(LADMH PUBLIC ...)` to include the CImg directory, wherever it is.
